src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -1749,3 +1784,292 @@ bool HNSWIndex<DataType, DistType>::preferAdHocSearch(size_t subsetSize, size_t
         res ? (initial_check ? HYBRID_ADHOC_BF : HYBRID_BATCHES_TO_ADHOC_BF) : HYBRID_BATCHES;
     return res;
 }
+
+#ifdef BUILD_TESTS
+template <typename DataType, typename DistType>
+void HNSWIndex<DataType, DistType>::saveIndexIMP(std::ofstream &output,
+                                                 EncodingVersion version) const {
+    // We already checked in the serializer that this is a valid version number.
+    // Now checking the version number to decide which data to write.
+    if (version != EncodingVersion_V1) {
+        saveIndexFields_v2(output);
+    }
+
+    this->saveIndexFields(output);
+    this->saveGraph(output);
+}
+
+template <typename DataType, typename DistType>
+void HNSWIndex<DataType, DistType>::loadIndexIMP(std::ifstream &input, EncodingVersion version) {
+    // We already checked in the serializer that this is a valid version number.
+    if (version != EncodingVersion_V1) {
+        // This data is only serialized from V2 up.
+        VecSimAlgo algo = VecSimAlgo_BF;
+        readBinaryPOD(input, algo);
+        if (algo != VecSimAlgo_HNSWLIB) {
```

Comment:
cant do `VecSimAlgo algo;` because some architectures identify this as a condition on an initialized value and fail valgrind tests